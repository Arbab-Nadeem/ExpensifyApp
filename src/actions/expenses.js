// import { v4 as uuidv4 } from 'uuid';
import { get, push, ref, remove, update } from 'firebase/database';

// ADD_EXPENSE
import database from '../firebase/firebase';
const addExpense = (expense) => ({
	type: 'ADD_EXPENSE',
	expense,
});
// const database = getDatabase();
/* const addExpense = ({
	description = '',
	note = '',
	amount = 0,
	createdAt = 0,
} = {}) => ({
	type: 'ADD_EXPENSE',
	expense: {
		id: uuidv4(),
		description,
		note,
		amount,
		createdAt,
	},
});
 */
const startAddExpenses = (expenseData = {}) => {
	return (dispatch, getState) => {
		/* this function gets called internally by redux and it gets called with dispatch. This just gives us access to dispatch so we can use it inside of here after we're done doing whatever we're going to be writing some data to Firebase waiting for that data to correctly sync. Then we'll use dispatch to dispatch add expense, making sure the Redux store reflects those changes*/
		const uid = getState().auth.uid;

		const {
			description = '',
			note = '',
			amount = 0,
			createdAt = 0,
		} = expenseData;
		const expense = { description, amount, note, createdAt };
		return push(ref(database, `users/${uid}/expenses`), expense).then(
			(reference) => {
				dispatch(
					addExpense({
						id: reference.key,
						...expense,
					})
				);
			}
		);
	};
};
//REMOVE_EXPENSE
const removeExpense = ({ id } = {}) => ({
	type: 'REMOVE_EXPENSE',
	id,
});

const startRemoveExpense = ({ id } = {}) => {
	/* dispatch will be passed to this function by redux library */
	return (dispatch, getState) => {
		const uid = getState().auth.uid;
		return remove(ref(database, `users/${uid}/expenses/${id}`)).then(
			() => {
				dispatch(removeExpense({ id }));
			}
		);
	};
};
//EDIT_EXPENSE
const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates,
});

const startEditExpense = (id, updates) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;
		return update(
			ref(database, `users/${uid}/expenses/${id}`),
			updates
		).then(() => dispatch(editExpense(id, updates)));
	};
};

const setExpenses = (expenses) => ({
	type: 'SET_EXPENSES',
	expenses,
});
const startSetExpenses = () => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;
		return get(ref(database, `users/${uid}/expenses`)).then(
			(snapshot) => {
				const expenses = [];
				snapshot.forEach((childSnapshot) => {
					expenses.push({
						id: childSnapshot.key,
						...childSnapshot.val(),
					});
				});
				dispatch(setExpenses(expenses));
			}
		);
	};
};

export {
	setExpenses,
	startSetExpenses,
	addExpense,
	startAddExpenses,
	removeExpense,
	startRemoveExpense,
	editExpense,
	startEditExpense,
};
