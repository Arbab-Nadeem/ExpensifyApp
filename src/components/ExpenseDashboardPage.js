import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

// ExpenseDashboardPage
const ExpenseDashboardPage = () => (
	<div>
		<ExpenseListFilters />
		<ExpenseList />
	</div>
);

export default ExpenseDashboardPage;
