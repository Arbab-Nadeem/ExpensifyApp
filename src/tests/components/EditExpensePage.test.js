import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense, startRemoveExpense, wrapper, history;
beforeEach(() => {
	editExpense = jest.fn();
	startRemoveExpense = jest.fn();
	history = { push: jest.fn() };
	wrapper = shallow(
		<EditExpensePage
			startRemoveExpense={startRemoveExpense}
			editExpense={editExpense}
			history={history}
			expense={expenses[1]}
		/>
	);
});
test('Should render the edit expense page correctly', () => {
	expect(wrapper).toMatchSnapshot();
});

test('Should handle onSubmit', () => {
	wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
	expect(history.push).toHaveBeenLastCalledWith('/');
	expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});
test('Should handle onClick', () => {
	wrapper.find('button').simulate('click');
	expect(history.push).toHaveBeenLastCalledWith('/');
	expect(startRemoveExpense).toHaveBeenLastCalledWith({
		id: expenses[1].id,
	});
});
