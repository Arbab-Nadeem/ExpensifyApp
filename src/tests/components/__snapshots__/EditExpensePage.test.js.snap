// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the edit expense page correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 1000,
        "createdAt": -345600000,
        "description": "coffee",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
